controllers/spifilecontentrequest_controller.go

```diff
@@ -154,50 +133,53 @@ func (r *SPIFileContentRequestReconciler) Reconcile(ctx context.Context, req ctr
-		if sp.GetDownloadFileCapability() == nil {
-			r.updateFileRequestStatusError(ctx, &request, serviceprovider.FileDownloadNotSupportedError{})
-			return ctrl.Result{}, serviceprovider.FileDownloadNotSupportedError{}
-		}
-		request.Status.Phase = api.SPIFileContentRequestPhaseAwaitingBinding
-	} else if request.Status.Phase == api.SPIFileContentRequestPhaseDelivered {
+	if request.Status.Phase == api.SPIFileContentRequestPhaseDelivered {
 		// already injected, nothing to do
 		return ctrl.Result{}, nil
 	}
+	// find associated binding
+	var associatedBinding *api.SPIAccessTokenBinding = nil
+	bindingList := &api.SPIAccessTokenBindingList{}
+	if err := r.K8sClient.List(context.Background(), bindingList, client.InNamespace(request.GetNamespace()), client.MatchingLabels{linkedBindingLabel: request.Name}); err != nil {
+		spiFileContentRequestLog.Error(err, "Unable to fetch bindings list", "namespace", request.GetNamespace())
+		return ctrl.Result{}, err
+	}
+	if len(bindingList.Items) > 0 {
+		associatedBinding = &bindingList.Items[0]
+	}
 
-	if request.Status.LinkedBindingName == "" {
-		if request.Status.Phase == api.SPIFileContentRequestPhaseError {
-			//we failed even before binding was created. most probably URL is not supported, no reason to continue
-			return ctrl.Result{Requeue: false}, nil
+	if associatedBinding == nil {
+		// check if the URL is processable and create binding, otherwise fail fast
+		sp, _ := r.ServiceProviderFactory.FromRepoUrl(ctx, request.Spec.RepoUrl, request.Namespace)
+		if sp.GetDownloadFileCapability() == nil {
+			r.updateFileRequestStatusError(ctx, &request, serviceprovider.FileDownloadNotSupportedError{})
+			return ctrl.Result{}, serviceprovider.FileDownloadNotSupportedError{}
 		}
 		bindingName, err := r.createAndLinkBinding(ctx, &request)
 		if err != nil {
 			return ctrl.Result{}, fmt.Errorf("failed to create the object: %w", err)
 		}
+		request.Status.Phase = api.SPIFileContentRequestPhaseAwaitingBinding
 		lg = lg.WithValues("linked_binding", bindingName)
-	} else {
-		binding := &api.SPIAccessTokenBinding{}
-		if err := r.K8sClient.Get(ctx, client.ObjectKey{Name: request.Status.LinkedBindingName, Namespace: request.Namespace}, binding); err != nil {
-			if errors.IsNotFound(err) {
-				request.Status.LinkedBindingName = ""
-				r.updateFileRequestStatusError(ctx, &request, err)
-			}
+		if err := r.K8sClient.Status().Update(ctx, &request); err != nil {
+			log.FromContext(ctx).Error(err, "failed to update the file binding link", "error", err)
+			return reconcile.Result{Requeue: true}, nil
 		}
+		return ctrl.Result{Requeue: true}, nil //TODO: better to requeue after some interval
+	} else {
 		//binding not yet fully ready to work with, let's try next time
-		if binding.Status.Phase == "" || binding.Status.UploadUrl == "" {
-			return reconcile.Result{}, nil
+		if associatedBinding.Status.Phase == "" || associatedBinding.Status.UploadUrl == "" {
+			return reconcile.Result{Requeue: true}, nil
```

Comment:
This creates a sort of busy loop until the associated binding is reconciled. Wouldn't it be better to reintroduce the watch on the bindings (and maybe rework it with the newly introduced label)?